File: src/main/java/sharev/domain/card/controller/CardController.java
package sharev.domain.card.controller;

import jakarta.validation.Valid;
import java.time.LocalDateTime;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import sharev.domain.account.entity.Account;
import sharev.domain.card.dto.request.RequestUpdateCardInfoDto;
import sharev.domain.card.dto.response.ResponseCardDto;
import sharev.domain.card.dto.response.ResponseUpdateCardInfoDto;
import sharev.domain.card.service.CardService;

@RestController
@RequestMapping("/gathering/{gatheringId}/cards")
@RequiredArgsConstructor
public class CardController {

    private final CardService cardService;

    @PostMapping
    public ResponseEntity<Void> join(@PathVariable UUID gatheringId,
                                     @AuthenticationPrincipal Account account) {
        cardService.join(gatheringId, account.getId());

        return ResponseEntity.status(HttpStatus.CREATED)
                .build();
    }

    @PatchMapping
    public ResponseEntity<ResponseUpdateCardInfoDto> updateInfo(@PathVariable UUID gatheringId,
                                                                @AuthenticationPrincipal Account account,
                                                                @Valid @RequestBody RequestUpdateCardInfoDto requestUpdateCardInfoDto) {
        ResponseUpdateCardInfoDto responseUpdateCardInfoDto = cardService.updateInfo(
                gatheringId, account.getId(),
                requestUpdateCardInfoDto.version(),
                requestUpdateCardInfoDto.introductionText());

        return ResponseEntity.ok(responseUpdateCardInfoDto);
    }

    @PreAuthorize("@cardService.hasCompletedCard(authentication.principal, #gatheringId)")
    @GetMapping
    public ResponseEntity<Page<ResponseCardDto>> getAllCard(@PathVariable UUID gatheringId,
                                                            @AuthenticationPrincipal Account account,
                                                            @RequestParam("snapshotTime") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime snapshotTime,
                                                            Pageable pageable) {

        return ResponseEntity.ok(cardService.getAllCard(gatheringId, account.getId(), snapshotTime, pageable));
    }

    @PreAuthorize("@cardService.hasCompletedCard(authentication.principal, #gatheringId)")
    @GetMapping("/me")
    public ResponseEntity<ResponseCardDto> getMyCard(@PathVariable UUID gatheringId,
                                                     @AuthenticationPrincipal Account account) {
        ResponseCardDto myCard = cardService.getMyCard(gatheringId, account.getId());

        return ResponseEntity.ok(myCard);
    }

    @PreAuthorize("@cardService.hasCompletedCard(authentication.principal, #gatheringId)")
    @GetMapping("by-pin/{pinNumber}")
    public ResponseEntity<ResponseCardDto> getCardByPinNumber(@PathVariable UUID gatheringId,
                                                              @AuthenticationPrincipal Account account,
                                                              @PathVariable Integer pinNumber) {
        ResponseCardDto participantCard =
                cardService.getCardByPinNumber(gatheringId, account.getId(), pinNumber);

        return ResponseEntity.ok(participantCard);
    }
}

File: src/main/kotlin/sharev/card/adapter/inbound/web/controller/CardController.kt
package sharev.card.adapter.inbound.web.controller

import jakarta.validation.Valid
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.format.annotation.DateTimeFormat
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*
import sharev.card.adapter.inbound.web.dto.request.UpdateCardIntroduceRequest
import sharev.card.adapter.inbound.web.dto.response.CardResponse
import sharev.card.adapter.inbound.web.dto.response.JoinCardResponse
import sharev.card.adapter.inbound.web.dto.response.MyPinNumberResponse
import sharev.card.adapter.inbound.web.dto.response.UpdateCardIntroduceResponse
import sharev.card.adapter.inbound.web.mapper.toCommand
import sharev.card.adapter.inbound.web.mapper.toMyPinNumberResponse
import sharev.card.adapter.inbound.web.mapper.toResponse
import sharev.card.application.port.inbound.command.*
import sharev.card.application.port.inbound.usecase.*
import sharev.common.adapter.inbound.security.model.AccountPrincipal
import java.time.LocalDateTime
import java.util.*

@RestController
@RequestMapping("/gatherings/{gatheringId}/cards")
class CardController(
    private val joinCardUseCase: JoinCardUseCase,
    private val updateCardInfoUseCase: UpdateCardInfoUseCase,
    private val getCardByPinNumberUseCase: GetCardByPinNumberUseCase,
    private val getMyCardUseCase: GetMyCardUseCase,
    private val getMyPinNumberUseCase: GetMyPinNumberUseCase,
    private val getAllCardsUseCase: GetAllCardsUseCase,
) {
    @PostMapping
    fun join(
        @PathVariable gatheringId: UUID,
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
    ): ResponseEntity<JoinCardResponse> {
        val response = joinCardUseCase.join(
            JoinCardCommand(gatheringId, accountPrincipal.id)
        ).toResponse()

        return ResponseEntity.status(HttpStatus.CREATED)
            .body(response)
    }

    @PatchMapping
    fun updateIntroduce(
        @PathVariable gatheringId: UUID,
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
        @Valid @RequestBody request: UpdateCardIntroduceRequest,
    ): ResponseEntity<UpdateCardIntroduceResponse> {
        val response = updateCardInfoUseCase.updateIntroduce(
            request.toCommand(gatheringId, accountPrincipal.id)
        ).toResponse()

        return ResponseEntity.ok(response)
    }

    @GetMapping
    fun getAllCards(
        @PathVariable gatheringId: UUID,
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
        @RequestParam("snapshotTime") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) snapshotTime: LocalDateTime,
        pageable: Pageable,
    ): ResponseEntity<Page<CardResponse>> {
        return ResponseEntity.ok(
            getAllCardsUseCase.getAllCards(
                GetAllCardsCommand(gatheringId, accountPrincipal.id, snapshotTime, pageable)
            ).map { it.toResponse() }
        )
    }

    @GetMapping("/me")
    fun getMyCard(
        @PathVariable gatheringId: UUID,
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
    ): ResponseEntity<CardResponse> {
        return ResponseEntity.ok(
            getMyCardUseCase.getMyCard(
                GetMyCardCommand(gatheringId, accountPrincipal.id)
            ).toResponse()
        )
    }

    @GetMapping("/me/pin")
    fun getMyPinNumber(
        @PathVariable gatheringId: UUID,
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
    ): ResponseEntity<MyPinNumberResponse> {
        return ResponseEntity.ok(
            getMyPinNumberUseCase.getMyPinNumber(
                GetMyPinNumberCommand(gatheringId, accountPrincipal.id)
            ).toMyPinNumberResponse()
        )
    }

    @GetMapping("by-pin/{pinNumber}")
    fun getCardByPinNumber(
        @PathVariable gatheringId: UUID,
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
        @PathVariable pinNumber: Int,
    ): ResponseEntity<CardResponse> {
        return ResponseEntity.ok(
            getCardByPinNumberUseCase.getCardByPinNumber(
                GetCardByPinNumberCommand(gatheringId, accountPrincipal.id, pinNumber)
            ).toResponse()
        )
    }
}
